### PR TITLE
#2068 added At first download, ask to choose beetween internal and ex…

### LINKED ISSUE #2068: At first download, ask to choose beetween "internal" and "external"
This is an idea to avoid users to download to the "wrong" place because for the beginners it is not obvious what are all the options the app provides.

### PR DESCRIPTION
**Pull request by cdhiraj40**
Fixes #2068 

**Screenshots** 

https://user-images.githubusercontent.com/75211982/144044600-89777444-c1b1-4119-a607-d72df28ada2c.mp4

### CODE REVIEW

File: core/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -127,8 +127,10 @@
   <string name="tts_pause">pause</string>
   <string name="tts_resume">resume</string>
   <string name="stop">stop</string>
+  <string name="configure_storage_option_title">Download book to internal storage?</string>
```
Comment (gouri-panda) [inline core/src/main/res/values/strings.xml:130]: Is the string value appropriate?

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt [hunk 2]
```diff
@@ -256,17 +258,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
...
-        })
+      else -> if (sharedPreferenceUtil.showStorageOption) {
+        showStorageConfigureDialog()
+      } else {
+        availableSpaceCalculator.hasAvailableSpaceFor(item,
+          { downloadFile(item.book) },
+          {
+            libraryList.snack(
+              getString(R.string.download_no_space) +
+                "\n" + getString(R.string.space_available) + " " +
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:269]: Maybe concatenate strings would be a better idea?

File: core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt [hunk 3]
```diff
@@ -172,6 +178,7 @@ class SharedPreferenceUtil @Inject constructor(val context: Context) {
     private const val PREF_NEW_TAB_BACKGROUND = "pref_newtab_background"
     private const val PREF_STORAGE_TITLE = "pref_selected_title"
     private const val PREF_EXTERNAL_LINK_POPUP = "pref_external_link_popup"
+    private const val PREF_SHOW_STORAGE_OPTION = "show_storgae_option"
```
Comment (gouri-panda) [inline core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt:181]: [nit] "pref_storage_option"

File: core/src/main/res/values/strings.xml (hunk 1 above)
Comment (4shutosh) [reply core/src/main/res/values/strings.xml:130]: I suggest something like"Choose the storage location"

Comment (cdhiraj40) [reply core/src/main/res/values/strings.xml:130]: if we choose "Choose the storage location" then the buttons having strings
will be wrong,
"configure" "yes"
 I followed what @kelson42 said in #2624 
 here->https://github.com/kiwix/kiwix-android/pull/2624#issuecomment-813025562
 I would sure change, but do look once at the comment :)

Comment (4shutosh) [reply core/src/main/res/values/strings.xml:130]: Oh okay, my bad

Comment (gouri-panda) [reply core/src/main/res/values/strings.xml:130]: I'm sorry. My bad I forget that comment.

Comment (cdhiraj40) [reply core/src/main/res/values/strings.xml:130]: its total fine :) @gouri-panda @4shutosh

**Comment by cdhiraj40** [comment]
@gouri-panda i have done the changes :)

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt [hunk 4]
```diff
@@ -256,17 +258,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
...
+      else -> if (sharedPreferenceUtil.showStorageOption) {
+        showStorageConfigureDialog()
+      } else {
+        availableSpaceCalculator.hasAvailableSpaceFor(item,
+          { downloadFile(item.book) },
+          {
+            libraryList.snack(
+              "${getString(R.string.download_no_space)} " +
+                "\n  ${getString(R.string.space_available)}" +
+                it,
```
Comment (s-ayush2903) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]:
```suggestion
              "${getString(R.string.download_no_space)}\n${getString(R.string.space_available)} $it",
```
We don't use "+" for joining Strings in Kotlin, this is called _concatenation_, instead interpolation is used to make the code more clean and legible
You can learn about string interpolation from here: https://kotlinlang.org/docs/idioms.html#string-interpolation

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: oh yes, I do remember doing this but when I committed this I got the error 
``Exceeded max line length`` so I had to take them to the next line which automatically added '+'
any suggestion to resolve this and use your suggested code instead?

Comment (s-ayush2903) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: Cool, you can fix lint errors via hitting `Ctrl + Alt + L`, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings

**Comment by gouri-panda** [comment]
>@gouri-panda I have done the changes :)

@cdhiraj40  Sorry! My laptop kind of broke right now. but I'll review your PR right after it fixes

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt (hunk 4 above)
Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: sure will commit these changes soon :)

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: >Cool, you can fix lint errors via hitting Ctrl + Alt + L, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings
I tried it but using that doesn't give proper output
image

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: >Cool, you can fix lint errors via hitting Ctrl + Alt + L, and in this case can use kotlin multiline strings:
https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings

@s-ayush2903 I tried it but using that doesn't give proper output
image

Comment (s-ayush2903) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: You're supposed to figure that on your own, its not difficult :stuck_out_tongue: , you can try and test on scratches in android studio.
Can refer: https://www.jetbrains.com/help/idea/scratches.html & https://medium.com/@shivam.gosavi340_58315/productivity-hack-androidstudio-kotlin-scratch-file-e17cfde152c6

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:270]: okay sure :)

Comment (s-ayush2903) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt]: Good work. Btw, can you please put a space on the rhs of colon, like it'll be easier to read if written as:
`Space available: xxx` instead `Space available:xxx`

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt]: image
done, thanks for suggestion that does makes it easier to read :)

**Comment by cdhiraj40** [comment]
please review @gouri-panda

**Comment by cdhiraj40** [comment]
please review @gouri-panda

**Comment by kelson42** [comment]
@gouri-panda I think @cdhiraj40 is expecting your review on this PR since 4 monthes :( Could you please have a look? Would love to bring this in next release.

**Comment by gouri-panda** [comment]
@cdhiraj40 @kelson42 sorry! I'll review this PR asap.

**Review by gouri-panda — COMMENTED** [review]
@cdhiraj40 We have changed our option to store the zim files a little bit in #2803. Could you update your PR accordingly?

**Comment by kelson42** [comment]
@cdhiraj40 Coukd you please rebase your PR (and resolve conflicts)? Would be great to integrate your fix in 3.6.0.

**Comment by cdhiraj40** [comment]
> @cdhiraj40 Coukd you please rebase your PR (and resolve conflicts)? Would be great to integrate your fix in 3.6.0.

yep sure, done!

**Comment by cdhiraj40** [comment]
> @cdhiraj40 We have changed our option to store the zim files a little bit in #2803. Could you update your PR accordingly?

I am not sure what is changed via #2803 that has relations with this. Could you explain to me what has been changed? Its been a long time since I was active in this project 😅

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!

hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

**Comment by cdhiraj40** [comment]
> > @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!
> 
> hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

Yes I understood that but I was wondering why we need to make any changes here as this PR adds a dialog box asking where to store the zim right?

**Comment by cdhiraj40** [comment]
> > @MohitMaliFtechiz Can you help @cdhiraj40 ? Thanks!
> 
> hi @cdhiraj40 , We have added file picker so that for in non play store release version user can select zim file from any storage by its own to read. We have added floating action button in locallibraryfragment.

Yes, I understood that but I was wondering why we need to make any changes here as this PR adds a dialog box asking where to store the zim right? 
cc @gouri-panda

**Comment by cdhiraj40** [comment]
@gouri-panda I think that has already been done!

**Comment by cdhiraj40** [comment]
@gouri-panda I think this is now good to go? Do let me know!

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt [hunk 5]
```diff
@@ -433,19 +433,22 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
...
-              getString(R.string.download_no_space) +
-                "\n" + getString(R.string.space_available) + " " +
-                it,
-              R.string.download_change_storage,
-              ::showStorageSelectDialog
-            )
-          }
-        )
+        else -> if (sharedPreferenceUtil.showStorageOption) {
+          showStorageConfigureDialog()
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:437]: hi @cdhiraj40 ,

So now we have two version of build specifically for android 11 and above
a) Play-Store version:- App specific without ```MANAGE_EXTERNAL_PERMISSION```
b) F-Droid version:- App specific with ```MANAGE_EXTERNAL_PERMISSION```

So you have place https://github.com/kiwix/kiwix-android/blob/52f7002ba65f080ac2330b0d4b018ca73afe7d13/core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt#L74 and ```Build.VERSION.SDK_INT >= Build.VERSION_CODES.R``` check before showing dialog.

Comment (cdhiraj40) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt:437]: ah got it, okay will do!

**Comment by kelson42** [comment]
@cdhiraj40 Stop merging main branch in your feature branch. Use rebase instead.

**Comment by cdhiraj40** [comment]
> @cdhiraj40 Stop merging main branch in your feature branch. Use rebase instead.

Yes as discussed earlier, I will do that. Its an old commit!

**Comment by kelson42** [comment]
@cdhiraj40 Ok, sorry, though it was recent.

**Comment by cdhiraj40** [comment]
> @cdhiraj40 Ok, sorry, though it was recent.

no worries, I will try it out and see if that works! ignore if you see something wrong here, I will make it right haha!

**Comment by kelson42** [comment]
@MohitMaliFtechiz We need to move quickly forward now. Please feel free superseed this PR so we can at the end merge.

**Review by MohitMaliFtechiz — APPROVED** [review]
Tested and Aproved.

**Pull request merged**